File: clients/ios/AwareChat-iOS/AwareChat-iOS/Core/Persistence/Messages/MessageLocalRepository.swift
import Foundation

nonisolated enum MessageDirection: String, Sendable { case outgoing, incoming }
nonisolated enum MessageState: String, Sendable { case pendingToSend, sending, sent, failed }

nonisolated struct LocalMessage: Equatable, Sendable, Identifiable {
  let message: MessageDTO
  let direction: MessageDirection
  let state: MessageState?
  let receivedAt: Date?
  var id: UUID { message.messageId }
  var displayTimestamp: Date { receivedAt ?? message.clientCreatedAt }

  /// Local acceptance metadata must never be echoed in a send_message request.
  func outgoingPayload() throws -> MessageDTO {
    guard direction == .outgoing else { throw PersistenceError.invalidData }
    return try MessageDTO(
      messageId: message.messageId, text: message.text, senderId: message.senderId,
      receiverId: message.receiverId, clientCreatedAt: message.clientCreatedAt,
      clientSequence: message.clientSequence
    )
  }
}

@MainActor
protocol MessageLocalRepository: Sendable {
  func message(id: UUID) throws -> LocalMessage?
  func messages(conversationId: String) throws -> [LocalMessage]
  func pendingMessages() throws -> [LocalMessage]
  @discardableResult func enqueue(text: String, receiverId: UUID, createdAt: Date) throws -> LocalMessage
  @discardableResult func persistIncoming(_ message: MessageDTO, receivedAt: Date) throws -> LocalMessage
  func markSending(id: UUID) throws
  func markAccepted(id: UUID, serverReceivedAt: Date) throws
  func markRejected(id: UUID, retryable: Bool) throws
  func recoverInterruptedSends() throws
  func observeMessages(conversationId: String) -> AsyncThrowingStream<[LocalMessage], any Error>
}

File: clients/android/AwareChat-Android/app/src/main/java/com/example/awarechat_android/core/persistence/messages/MessageLocalRepository.kt
package com.example.awarechat_android.core.persistence.messages

import com.example.awarechat_android.core.persistence.database.PersistenceException
import com.example.awarechat_android.core.persistence.database.PersistenceFailure
import com.example.awarechat_android.core.protocol.MessageDto
import java.time.Instant
import java.util.UUID
import kotlinx.coroutines.flow.Flow

enum class MessageDirection {
    OUTGOING,
    INCOMING,
}

enum class MessageState {
    PENDING_TO_SEND,
    SENDING,
    SENT,
    FAILED,
}

data class LocalMessage(
    val message: MessageDto,
    val direction: MessageDirection,
    val state: MessageState?,
    val receivedAt: Instant?,
) {
    val displayTimestamp: Instant
        get() = receivedAt ?: message.clientCreatedAt

    fun outgoingPayload(): MessageDto {
        if (direction != MessageDirection.OUTGOING) {
            throw PersistenceException(PersistenceFailure.INVALID_DATA)
        }
        return message.copy(serverReceivedAt = null)
    }
}

interface MessageLocalRepository {
    suspend fun message(id: UUID): LocalMessage?

    suspend fun messages(conversationId: String): List<LocalMessage>

    suspend fun pendingMessages(): List<LocalMessage>

    suspend fun enqueue(text: String, receiverId: UUID, createdAt: Instant): LocalMessage

    suspend fun persistIncoming(message: MessageDto, receivedAt: Instant): LocalMessage

    suspend fun markSending(id: UUID)

    suspend fun markAccepted(id: UUID, serverReceivedAt: Instant)

    suspend fun markRejected(id: UUID, retryable: Boolean)

    suspend fun recoverInterruptedSends()

    fun observeMessages(conversationId: String): Flow<List<LocalMessage>>
}
